File: bots/apis/codemarket.py
import json
import requests

from typing import Optional

BASE_URL = 'http://localhost:8000'

"""CodeMarket Python API

This module is for interacting with the CodeMarket server using python. Do not
send any sensitive information through these connections.
"""

def base_api_get(url: str) -> dict:
    """Base API GET request to get data from the market
    
    Args:
        url   (str):    The url to send the request to

    Returns:
        dict:   Contains the response content of the call
    """
    r = requests.get(BASE_URL + url)
    return json.loads(r.content.decode())

def base_api_post(url: str, data: dict) -> dict:
    """Base API POST request to change data in the market
    
    Args:
        url   (str):    The url to send the request to
        data (dict):    The data being sent to the endpoint

    Returns:
        dict:   Contains the response content of the call
    """
    headers = {'content-type': 'application/x-www-form-urlencoded'}
    r = requests.post(BASE_URL + url, data=data, headers=headers)
    return json.loads(r.content.decode())

def get_ledger_state(uuid: str) -> dict:
    """Get the current ledger state
    
    Args:
        uuid (str):     Your UUID to verify the ledger state request

    Returns:
        dict:   Contains the current ledger state
    """
    payload = { 'uuid': uuid }
    return base_api_post('/api/ledger_state', data=payload)

def get_vendor_names() -> dict:
    """Get a list of registered vendor names

    Returns:
        dict:   Contains a list of registered vendor names
    """
    return base_api_get('/api/vendor_names')

def get_vendor_urls() -> dict:
    """Get a list of vendor urls

    Returns:
        dict:   Contains a list of vendor urls
    """
    return base_api_get('/api/vendor_urls')

def purchase(item: str, count: int, frm: str, to: str) -> dict:
    """Purchase an item FROM the vendor TO the buyer

    Args:
        item  (str):    The name of the item
        count (int):    Amount to purchase
        frm   (str):    Name of the vendor to purchase from
        to    (str):    Your UUID to verify the purchase
    
    Returns:
        dict:   Contains purchase receipt or errors
    """
    payload = {
        'item': item,
        'count': count,
        'from': frm,
        'to': to
    }
    return base_api_post('/api/purchase', data=payload)

def register_vendor(vendor_name: str, vendor_url: Optional[str] = '') -> dict:
    """Register a new vendor with the market

    Args:
        vendor_name          (str):     The name of the vendor
        vendor_url (Optional[str]):     The url of the vendor
    
    Returns:
        dict:   Contains new vendor's UUID or errors
    """
    payload = { 
        'vendor_name': vendor_name,
        'vendor_url': vendor_url
    }
    return base_api_post('/register', data=payload)

def stock(item: str, price: float, stock: int, uuid: str) -> dict:
    """Stock/store item within your a shop

    Args:
        item    (str):  The name of the item
        price (float):  Amount to price of item
        stock   (int):  The amount of items to go from the store to the stock,
                        negative values will move items from the stock to the
                        store
        uuid    (str):  Your UUID to verify the stock request
    
    Returns:
        dict:   Contains stocking receipt or errors
    """
    payload = {
        'name': item,
        'price': price,
        'stock': stock,
        'uuid': uuid
    }
    return base_api_post('/api/stock', data=payload)

File: bots/apis/codemarket.js
/** 
 * CodeMarket JavaScript API
 * 
 * This module is for interacting with the CodeMarket server using javascript.
 * Do not send any sensitive information through these connections.
 */

const BASE_URL = 'http://[::1]:8000'; // JS localhost url

/**
 * Base API GET request to get data from the market
 * @param {string} url - The request extension
 * @returns {object} The return object
 */
exports.base_api_get = async function(url) {
    const got = require('got');

    try {
        const response = await got(BASE_URL.concat(url));
        return JSON.parse(response.body);
    } catch (error) {
        return JSON.parse(error.response.body);
    }
};

/**
 * Base API POST request to change data in the market
 * @param {string} url - The request extension
 * @param {object} data - The data being sent to the endpoint
 * @returns {object} The return object
 */
exports.base_api_post = async function(url, data) {
    const got = require('got');

    try {
        const response = await got.post(BASE_URL.concat(url), {form: data});
        return JSON.parse(response.body);
    } catch (error) {
        return JSON.parse(error.response.body);
    }
};

/**
 * Get the current ledger state
 * @param {string} uuid - Your UUID for verification
 * @returns {object} The current ledger state
 */
exports.get_ledger_state = async function(uuid) {
    payload = { uuid };
    return await exports.base_api_post('/api/ledger_state', payload);
};

/**
 * Get a list of registered vendors
 * @returns {object} A list of currently registered vendors
 */
exports.get_vendor_names = async function() {
    return await exports.base_api_get('/api/vendor_names');
};

/**
 * Get a list of vendor urls
 * @returns {object} A list of urls for all registered vendors
 */
exports.get_vendor_urls = async function() {
    return await exports.base_api_get('/api/vendor_urls');
};

/**
 * Purchase an item FROM the vendor TO the buyer
 * @param {string} item - The name of the item
 * @param {integer} count - Amount to purchase
 * @param {string} frm - Name of the vendor to purchase from
 * @param {string} to - Your UUID to verify the purchase
 * @returns {object} Contains receipt or errors
 */
exports.purchase = async function(item, count, frm, to) {
    payload = { item, count, frm, to };
    return await exports.base_api_post('/api/purchase', payload);
};

/**
 * Register a new vendor with the market
 * @param {string} vendor_name - The name of the vendor
 * @param {string} vendor_url - Optional custom url for vendor
 * @returns {object} Contains new UUID or errors
 */
exports.register_vendor = async function(vendor_name, vendor_url = '') {
    payload = { vendor_name, vendor_url };
    return await exports.base_api_post('/register', payload);
};

/**
 * 
 * @param {string} name - The name of the item
 * @param {float} price - Amount to set price of item
 * @param {integer} stock - The amount of items to go from the store to the 
 *                          stock, negative values will move items from the 
 *                          stock to the store
 * @param {string} uuid - Your UUID to verify the stock request
 * @returns {object} Contains stocking receipt or errors
 */
exports.stock = async function(name, price, stock, uuid) {
    payload = {name, price, stock, uuid};
    return await exports.base_api_post('/api/stock', payload);
};